tests/functional/src/main/java/org/mule/functional/classloading/isolation/classloader/MuleClassLoaderRunnerFactory.java

```diff
@@ -0,0 +1,265 @@
+     * @return a {@link CompositeClassLoader} that represents the plugin class loaders.
+     */
+    private ClassLoader createPluginClassLoaders(ClassLoader parent, ClassLoaderLookupPolicy childClassLoaderLookupPolicy, ArtifactUrlClassification artifactUrlClassification, List<ArtifactClassLoader> extensionPluginsArtifactClassLoaders)
+    {
+        final List<ClassLoader> pluginClassLoaders = new ArrayList<>();
+        pluginClassLoaders.add(new MuleArtifactClassLoader("sharedLibs", new URL[0], parent, childClassLoaderLookupPolicy));
+
+        for (PluginUrlClassification pluginUrlClassification : artifactUrlClassification.getPluginClassificationURLs())
+        {
+            logClassLoaderUrls("PLUGIN (" + pluginUrlClassification.getName() + ")", pluginUrlClassification.getUrls());
+            MuleArtifactClassLoader pluginCL = new MuleArtifactClassLoader(pluginUrlClassification.getName(), pluginUrlClassification.getUrls().toArray(new URL[0]), parent, childClassLoaderLookupPolicy);
+
+            Collection<String> exportedPackages;
+            Collection<String> exportedResources;
+            if (pluginUrlClassification instanceof ExtensionPluginUrlClassification)
+            {
+                extensionPluginsArtifactClassLoaders.add(pluginCL);
+
+                URL manifestUrl = pluginCL.findResource("META-INF/" + EXTENSION_MANIFEST_FILE_NAME);
+                ExtensionManifest extensionManifest = extensionManager.parseExtensionManifestXml(manifestUrl);
+
+                exportedPackages = extensionManifest.getExportedPackages();
+                exportedResources = extensionManifest.getExportedResources();
+            }
+            else
+            {
+                ClassLoader pluginArtifactClassLoaderToDiscoverModules = new URLClassLoader(pluginUrlClassification.getUrls().toArray(new URL[0]), null);
+                List<MuleModule> modules = withContextClassLoader(pluginArtifactClassLoaderToDiscoverModules, () -> new ClasspathModuleDiscoverer(pluginArtifactClassLoaderToDiscoverModules).discover());
+                MuleModule module = validatePluginModule(pluginUrlClassification.getName(), modules);
+
+                exportedPackages = module.getExportedPackages();
+                exportedResources = module.getExportedPackages();
+            }
+            ArtifactClassLoaderFilter filter = artifactClassLoaderFilterFactory.create(exportedPackages.stream().collect(Collectors.joining(", ")),
+                                                                                       exportedResources.stream().collect(Collectors.joining(", ")));
+            pluginClassLoaders.add(new FilteringArtifactClassLoader(pluginCL, filter));
+        }
+        return new CompositeClassLoader(parent, pluginClassLoaders, childClassLoaderLookupPolicy);
+    }
+
+    /**
+     * Validates that only one module should be discovered.
```

Comment:
"Validates that only one module should be discovered. A plugin cannot have inside another plugin that holds a {@code mule-module.properties} for the time being."

When I implemented this I thought that it will be not supported to include other plugins inside plugins, for instance you cannot write you own plugin that has as library mule-core. This may change later whenever we would add support for composing plugins or share things among plugins. For the time being I'm checking that only one mule-module.propertites has to be found in the plugin classpath (URLs).